src/node_order_optim/cellorder.cpp

```diff
@@ -0,0 +1,793 @@
+        int* perm = interleave_order(tid, nt.ncell, nt.end, nt._v_parent_index);
+        auto p = inverse_permute(perm, nt.end);
+#if 0
+        for (int i = 0; i < nt.end; ++i) {
+            int x = nt._v_parent_index[p[i]];
+            int par = x >= 0 ? perm[x] : -1;
+            printf("%2d <- %2d  parent=%2d\n", i, p[i], par);
+        }
+#endif
+        //        prnode("before perm", nt);
+        forward_permute<Node*>(nt._v_node, p, nt.end);
+        forward_permute<Node*>(nt._v_parent, p, nt.end);
+        forward_permute<int>(nt._v_parent_index, p, nt.end);
+        node_permute(nt._v_parent_index, nt.end, perm);
+        for (int i = 0; i < nt.end; ++i) {
+            nt._v_node[i]->v_node_index = i;
+        }
+        for (auto tml = nt.tml; tml; tml = tml->next) {
+            Memb_list* ml = tml->ml;
+            for (int i = 0; i < ml->nodecount; ++i) {
+                ml->nodeindices[i] = perm[ml->nodeindices[i]];
+            }
+            sort_ml(ml);  // all fields in increasing nodeindex order
+        }
+        //        prnode("after perm", nt);
+
+        delete[] perm;
+    }
+    //    printf("leave nrn_permute_node_order\n");
+}
+
+int* interleave_order(int ith, int ncell, int nnode, int* parent) {
+    // return if there are no nodes to permute
+    if (nnode <= 0)
+        return nullptr;
+
+    // ensure parent of root = -1
+    for (int i = 0; i < ncell; ++i) {
+        if (parent[i] == 0) {
+            parent[i] = -1;
+        }
+    }
+
+    int nwarp = 0, nstride = 0, *stride = nullptr, *firstnode = nullptr;
+    int *lastnode = nullptr, *cellsize = nullptr, *stridedispl = nullptr;
+
+    int* order = node_order(
+        ncell, nnode, parent, nwarp, nstride, stride, firstnode, lastnode, cellsize, stridedispl);
+
+    if (interleave_info) {
+        InterleaveInfo& ii = interleave_info[ith];
+        ii.nwarp = nwarp;
+        ii.nstride = nstride;
+        ii.stridedispl = stridedispl;
+        ii.stride = stride;
+        ii.firstnode = firstnode;
+        ii.lastnode = lastnode;
+        ii.cellsize = cellsize;
+        if (0 && ith == 0 && interleave_permute_type == 1) {
+            printf("ith=%d nstride=%d ncell=%d nnode=%d\n", ith, nstride, ncell, nnode);
+            for (int i = 0; i < ncell; ++i) {
+                printf("icell=%d cellsize=%d first=%d last=%d\n",
+                       i,
+                       cellsize[i],
+                       firstnode[i],
+                       lastnode[i]);
+            }
+            for (int i = 0; i < nstride; ++i) {
+                printf("istride=%d stride=%d\n", i, stride[i]);
+            }
+        }
+        if (ith == 0) {
+            // needed for print_quality[12] and done once here to save time
+            int* p = new int[nnode];
+            for (int i = 0; i < nnode; ++i) {
+                p[i] = parent[i];
+            }
+            permute_ptr(p, nnode, order);
+            node_permute(p, nnode, order);
```

Comment:
@nrnhines I started working on the refactoring of the permutation algorithm. (Sorry for the delay, I was busy with other stuff)
Could you explain to me what's the difference between `permute_ptr` and `node_permute` and why we call them consecutively here?

Reply:
p is an array of parent indices. I.e p[i] is the parent node index of the ith node. As such p has to be permuted in the same way and nodes. In addition the value of p[i] must be changed to the new index of the permuted parent node. In node_permute.cpp you can observe in the ```permute_ptr``` implementation that the p vector is permuted according to order.  And in ```node_permute``` the value of p[i] is changed to order[p[i]]